src/transformers/models/depth_pro/modeling_depth_pro.py

```diff
@@ -0,0 +1,1571 @@
+        hidden_state = (
+            image_encodings.last_hidden_state
+        )  # (scaled_images_num_patches[0], self.seq_len+1, config.hidden_size)
+
+        # b. reshape back to image like
+        image_features = reshape_feature(
+            hidden_state, self.out_size, self.out_size
+        )  # (scaled_images_num_patches[0], config.hidden_size, self.out_size, self.out_size)
+
+        # c. merge patches back together
+        image_features = merge(
+            image_features,
+            batch_size=B,
+            merge_out_size=self.out_size * 2 ** (0),
+        )  # (B, config.hidden_size, self.out_size*2**(self.n_scaled_images-1), self.out_size*2**(self.n_scaled_images-1))
+
+        # d. upsample
+        image_features = self.upsample_image(
+            image_features
+        )  # (B, config.scaled_images_feature_dims[0], self.out_size*2**1, self.out_size*2**1)
+
+        # STEP 7: apply fusion (global_features = image_features + scaled_images_features[0])
+        # fuses image_features with lowest resolution features as they are of same size
+        scaled_images_features[0] = torch.cat((scaled_images_features[0], image_features), dim=1)
+        scaled_images_features[0] = self.fuse_image_with_low_res(scaled_images_features[0])
+
+        # STEP 8: return these features in order of increasing size as what fusion expects
+        features = [
+            # (B, self.scaled_images_feature_dims[i], self.out_size*2**(i+1), self.out_size*2**(i+1))
+            *scaled_images_features,
+            # (B, config.intermediate_feature_dims[i], self.out_size*2**(self.n_scaled_images+i+1), self.out_size*2**(self.n_scaled_images+i+1))
+            *intermediate_features,
+        ]
+
+        # prepare last_hidden_state, hidden_states, attentions from patches to batches
+
+        last_hidden_state = patch_encodings.last_hidden_state
+        hidden_states = patch_encodings.hidden_states if output_hidden_states else None
+        attentions = patch_encodings.attentions if output_attentions else None
+
+        num_patches = sum(scaled_images_num_patches)
+        # [0, 3, 6], [1, 4, 7], [2, 5, 8] when num_patches=9 and B=3
+        indexes = torch.arange(num_patches).reshape(num_patches//B, -1).T
+        indexes = indexes.to(last_hidden_state.device)
+
+        last_hidden_state = last_hidden_state[indexes].mean(1)
+        if hidden_states is not None:
+            hidden_states = tuple([state[indexes].mean(1) for state in hidden_states])
+        if attentions is not None:
+            attentions = tuple([state[indexes].mean(1) for state in attentions])
+
+        if not return_dict:
+            return tuple(v for v in [last_hidden_state, features, hidden_states, attentions] if v is not None)
+
+        return DepthProOutput(
+            last_hidden_state=last_hidden_state,
+            features=features,
+            hidden_states=hidden_states,
+            attentions=attentions,
+        )
```

Comment:
@qubvel can you please suggest a solution here

patches are created and concatenated from each image in a batch, thus the input and output from `patch_encoder` is of length equal to the number of patches instead of the batch size.

however these patches can be converted back to batches, as I have done using `indexes` here, but the problem is with the test case `test_retain_grad_hidden_states_attentions`, which happens because the conversion from patch to batch is never back-propagated.

Reply:
Hey @geetu040, not sure I got the problem here, is the problem using indexes? Can you avoid it using the same operation directly on hidden states?

```python
last_hidden_state = last_hidden_state.reshape(num_patches//B, -1).transpose(0, 1).mean(1)
```

Reply:
the problem is not with just indexing but any kind of operation on `last_hidden_state` or `hidden_states` or `attentions`

this is because during back-propagation, gradients are collected for them, but when we apply any kind of operation, the resultant tensors produced, didnot contribute to the final loss, therefore no gradient is accumulated for them and that fails the test case.

Reply:
I can fix this if I return these states from `image_encoder` instead of `patch_encoder`

both are vision transformers used in encoder
- `patch_encoder`
  - takes the patches from images of different scales
  - input vector is of shape `num_patches, channels, patch_size, patch_size`
  - indexing and mean is applied to match output hidden states to length batch size, which creates a new computation graph
- `image_encoder`
  - takes original image scaled to patch_size, thus takes that single patch/image as input
  - input vector is of shape `B, channels, patch_size, patch_size`
  - doesnot require indexing and mean, because output hidden states are already of length batch size, no new computation graph is created

Reply:
I haven't delved deeply into architectural details yet, so I would recommend choosing hidden states based more on their potential usage rather than on test design. Choose the ones that you believe might be more suitable for building a head on top. Tests can be overridden with a clear explanation of why we made that decision.

Reply:
I would prefer `hidden_states[0].shape # (batch_size, num_patches, seq_len, hidden_size)`, one can flatten them if needed

Reply:
https://github.com/huggingface/transformers/blob/0415722bd6dd44f4b7d56d0cacf8cdd3f958cb41/src/transformers/models/depth_pro/modeling_depth_pro.py#L239-L242

this is how I now handle this: `attention_probs` are reshaped to created batches which are again reshaped to create the original `attention_probs`, the batched tensor is returned as `output_attentions` and the patched tensor is used for further computations.

this way the shape conversions, become a part of the computation graph of backpropagation and the gradients are retained and the output intermediate states look as we discussed above

Reply:
By the way, we should always consider that input has a batch dimension, so I'm not sure why `batch_size` might be `None`. However, maybe a more in-depth review is needed.

Reply:
`batch_size` is a parameter, I added in the `forward` pass of some layers, provided only in `patch_encoder`, otherwise it is `None` by default.

Reply:
further discussed here: https://github.com/huggingface/transformers/pull/34583#discussion_r1894173861